### DISCUSSION TITLE
The solution with the lowest memory usage

### DISCUSSION THREAD

**Discussion by asswclw** [valhalla/valhalla#4826] [Q&A]
I find this project incredibly impressive and I would like to utilize it as a library in an embedded environment. The goal is to confine the program's memory usage to within 8MB, and I am willing to tolerate a slower response time. Do you have any advice on how I can achieve this? I'm aware this could be a challenging task and I'm currently short of ideas. I genuinely appreciate your attention on this matter.

**Comment by ImreSamu** [comment]
Before anyone can provide more detailed answers, could you provide more information?

- What specific CPU architecture are you using in the embedded environment? (32-bit vs. 64-bit)
- Which features of Valhalla are essential for your application? (The fewer, the better the chances of meeting your memory requirements)
- Do you need elevation data? Are street names or specific road information necessary? Any other specific data requirements?
- What is the geographic coverage area for the routing? (A few streets? A small town? A country? A continent?)
- Are there any particular constraints on processing power or storage capacity?
- How often will routing calculations need to be performed?
- etc.

**Comment by kevinkreiser** [comment]
8mb of ram is bold ill give you that... for any kind of solution that trades ram usage for speed im going to have to assume you have plenty of storage, if thats not the case you are dead in the water. assuming that is the case, lets talk about where the memory usage comes from:

1. the graph itself
2. the algorithms, primarily the edge labels which track the graph expansion (search paths)

## graph memory usage

We have some options for limiting the amount ram the graph itself will use in our tile cache. You could make use of the hard limit in the LRU cache. the problem there is that some of the tiles will be bigger than 8mb, i know for certain that there is a tile in japan (very dense network) that is 25mb. its possible to have the graph building stages skip edges that are uninteresting to your use-case (bike only or pedestrian only or car only) this can help to trim the graph size, but if this isnt enough we'd have to start thinking about larger changes to the tiles and that simple isnt going to happen (though you could do it in a fork, its a ton of work though).

if your device supports memory mapping (os thing that lets you treat a file as if it were in ram) then you are in luck. you can take your whole maps tileset put it in a tar and configure valhalla that way. if you use this you'll use the least memory possible. at the moment i dont think you can use this mode with a hard limit lru tile cache but we could add that i guess

## algorithm memory usage

this is a hard one. we currently dont have a way to tightly control the ram use here. we have reservations that we can minimize so that after each computation all the ram is given back but while its being computed we dont limit the amount used. i can think of a few ways we could deal with this problem.

the simplest would be to swap out `std::unordered_map` and the like for disk-backed variants such as `stxxl::unordered_map`. we could do this at compile time with `ifdef`s and `typedef`s. this would be slow as heck but it would take the bulk of the algorithmic allocation.

the other option would be to pass custom allocators to our stl containers, presumably memory mapped files being the backing store. maybe thats how stxxl already works im not sure but at least the os could allow some of the file to be swapped into actual ram as its needed rather than always going to disk for any operation

**Reply by asswclw** [reply]
Thank you so much for your patience in responding! Yes, my external storage is quite spacious, but my internal memory is somewhat limited. I'll take your advice into consideration and give it a go. If there's any positive outcome, I'll make sure to come back and share it with you right away. Thanks again, and have a wonderful day!"

**Reply by kevinkreiser** [reply]
no worries, thanks for the interesting question!
